# Logistic Regression carbon, silicon, and nitrogen isotopes

In [1]:
# Imports
import joblib
import pandas as pd

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

## Read the CSV and Perform Basic Data Cleaning

In [2]:
C_Si_N = pd.read_csv('presolargrains_C_Si_N.csv')
C_Si_N.head()

Type  carbon_12_13  nitrogen_14_15  silicon_29_28  silicon_30_28
0    X         74.30           207.4         -685.0         -520.0
1    X       1581.00           116.0         -684.0         -490.0
2    X        234.75           187.2         -683.0         -501.0
3    X        455.00           140.0         -662.0         -770.0
4    X        823.00            44.0         -658.0         -234.0

## Create a Train Test Split

In [3]:
# Set `Type` for the y values
y = C_Si_N["Type"]
X = C_Si_N.drop(columns=["Type"])

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

In [4]:
X_train.head()

carbon_12_13  nitrogen_14_15  silicon_29_28  silicon_30_28
457         121.80          1412.0           -1.6           35.7
86          185.00            62.0         -332.0         -467.0
646          20.09           500.0           27.8           29.8
1240         50.10          2160.0          132.0          109.0
523         175.18          4954.0            8.0           78.0

## Pre-processing

In [5]:
# Scale the data using the MinMaxScaler
X_scaler = MinMaxScaler().fit(X_train)

X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Train the Model

In [6]:
model = LogisticRegression()
model.fit(X_train_scaled, y_train)

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [7]:
print(f"Training Data Score: {model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {model.score(X_test_scaled, y_test)}")

Training Data Score: 0.7005128205128205
Testing Data Score: 0.7116564417177914


## Hyperparameter Tuning

In [8]:
# Use `GridSearchCV` to tune the model's parameters
param_grid = {'C': [1, 5, 10],
              'penalty': ["l1", "l2"]}
grid = GridSearchCV(model, param_grid, verbose=3)

In [9]:
grid.fit(X_train_scaled, y_train)

/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:652: Warning: The least populated class in y has only 2 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class

Fitting 3 folds for each of 6 candidates, totalling 18 fits
[CV] C=1, penalty=l1 .................................................
[CV] ........ C=1, penalty=l1, score=0.7308868501529052, total=   0.0s
[CV] C=1, penalty=l1 .................................................
[CV] ........ C=1, penalty=l1, score=0.7400611620795107, total=   0.0s
[CV] C=1, penalty=l1 .................................................
[CV] ........ C=1, penalty=l1, score=0.7445482866043613, total=   0.0s
[CV] C=1, penalty=l2 .................................................
[CV] ........ C=1, penalty=l2, score=0.6605504587155964, total=   0.0s
[CV] C=1, penalty=l2 .................................................
[CV] ........ C=1, penalty=l2, score=0.6788990825688074, total=   0.0s
[CV] C=1, penalty=l2 .................................................
[CV] ........ C=1, penalty=l2, score=0.6853582554517134, total=   0.0s
[CV] C=5, penalty=l1 .................................................
[CV] ........ C=5

GridSearchCV(cv='warn', error_score='raise-deprecating',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'C': [1, 5, 10], 'penalty': ['l1', 'l2']},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=3)

In [11]:
# Print the best parameters and the best score
print(grid.best_params_)
print(grid.best_score_)

{'C': 10, 'penalty': 'l1'}
0.8441025641025641


## Save the Model

In [12]:
# Save fitted model to file
filename = 'LogisticRegression_C_Si_N.sav'
joblib.dump(grid, filename)

['LogisticRegression_C_Si_N.sav']

## Summary:

### Logistic Regression based on carbon, silicon, and nitrogen isotopes: 84.4% at C=10, penalty=11